docs/eBpfExtensions.md

```diff
@@ -248,16 +248,65 @@ the provider must free the per-client context passed in via `ProviderBindingCont
 
 ### 2.5 Invoking an eBPF program from Hook NPI Provider
 To invoke an eBPF program, the extension uses the dispatch table supplied by the Hook NPI client during attaching.
-There is only one function in the client dispatch table, which is of the following type:
+The client dispatch table contains the functions, with the following type prototypes:
 
 ```
 /**
- *  @brief This is the only mandatory function in the eBPF Hook NPI client dispatch table.
+ * @brief Invoke the eBPF program.
+ *
+ * @param[in] extension_client_binding_context The context provided by the extension client when the binding was created.
+ * @param[in,out] program_context The context for this invocation of the eBPF program.
+ * @param[out] result The result of the eBPF program.
+ *
+ * @retval EBPF_SUCCESS if successful or an appropriate error code.
+ * @retval EBPF_NO_MEMORY if memory allocation fails.
+ * @retval EBPF_EXTENSION_FAILED_TO_LOAD if required extension is not loaded.
  */
 typedef ebpf_result_t (*ebpf_program_invoke_function_t)(
-    _In_ const void* client_binding_context, _In_ const void* context, _Out_ uint32_t* result);
+    _In_ const void* extension_client_binding_context, _Inout_ void* program_context, _Out_ uint32_t* result);
 
+/**
+ * @brief Prepare the eBPF program for batch invocation.
+ *
+ * @param[in] extension_client_binding_context The context provided by the extension client when the binding was created.
+ * @param[in] state_size The size of the state to be allocated.
```

Comment:
I'd recommend referring to `ebpf_execution_context_state_t` here, too, in addition to the remarks below.

Reply:
You may also want to consider changing `state` to be a `ebpf_execution_context_state_t *` rather than pointer to void - that should provide a nice hint to developers and reduce compile-time errors.

Reply:
The intent of keeping this opaque is to permit backwards compatibility. Extension compiled against a newer version of eBPF core may increase the size of ebpf_execution_context_state_t, but the extension should still work for an older version of the eBPF core.

Reply:
Yup - as long as the struct itself is treated as opaque by the caller (the struct itself is not defined as an opaque blob of bytes) the compat story should work correctly for both forward and back compat, right?

Back compat with older eBPF core: size parameter is greater or equal to the required size, eBPF is happy.
Forward compat with newer eBPF core: size parameter may be smaller than the required size, eBPF is unhappy and gracefully returns an error code.

Actually, this raises a question I had not considered in WTC, though it is dealt with in XDP: should the extension fall back to the non-batched API if the batched enter routine returns failure?

Reply:
Oh, and actually I removed that code in XDP, but it looked like this:

```
   EbpfContext = TryBatchEnter(...);

    if (EbpfContext != NULL) {
        ebpf_program_batch_invoke_function_t EbpfInvokeProgram =
            EbpfExtensionClientGetProgramDispatch(Client)->ebpf_program_batch_invoke_function;
        EbpfResult = EbpfInvokeProgram(ClientBindingContext, &XdpMd.Base, &Result, EbpfContext);
    } else {
        ebpf_program_invoke_function_t EbpfInvokeProgram =
            EbpfExtensionClientGetProgramDispatch(Client)->ebpf_program_invoke_function;
        EbpfResult = EbpfInvokeProgram(ClientBindingContext, &XdpMd.Base, &Result);
    }
```

Reply:
> Yup - as long as the struct itself is treated as opaque by the caller (the struct itself is not defined as an opaque blob of bytes) the compat story should work correctly for both forward and back compat, right?
> 
> Back compat with older eBPF core: size parameter is greater or equal to the required size, eBPF is happy. Forward compat with newer eBPF core: size parameter may be smaller than the required size, eBPF is unhappy and gracefully returns an error code.
> 
> Actually, this raises a question I had not considered in WTC, though it is dealt with in XDP: should the extension fall back to the non-batched API if the batched enter routine returns failure?

Actually, eBPF core *could* switch its behavior based on the size of the state. I.e. if it gets a state size == old size, fall back to old behavior, if it gets the new size, use the new behavior.

Reply:
> Oh, and actually I removed that code in XDP, but it looked like this:
> 
> ```
>    EbpfContext = TryBatchEnter(...);
> 
>     if (EbpfContext != NULL) {
>         ebpf_program_batch_invoke_function_t EbpfInvokeProgram =
>             EbpfExtensionClientGetProgramDispatch(Client)->ebpf_program_batch_invoke_function;
>         EbpfResult = EbpfInvokeProgram(ClientBindingContext, &XdpMd.Base, &Result, EbpfContext);
>     } else {
>         ebpf_program_invoke_function_t EbpfInvokeProgram =
>             EbpfExtensionClientGetProgramDispatch(Client)->ebpf_program_invoke_function;
>         EbpfResult = EbpfInvokeProgram(ClientBindingContext, &XdpMd.Base, &Result);
>     }
> ```

I am not sure there is any value in the fall back. Possible conditions in which it could fail:
1) Caller is passing a totally unsupported state size -> Caller error, caller should fix it.
2) System is under low memory pressure or dependent extension isn't loaded -> Both batch and non-batch will fail.